{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5fac74c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import cobra\n",
    "from tqdm import tqdm\n",
    "import csv\n",
    "import re\n",
    "import urllib\n",
    "import pickle\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65de1971",
   "metadata": {},
   "source": [
    "### DNA features "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3df5cb8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_features_10(x):\n",
    "    Cmodel = cobra.io.read_sbml_model('/data1/xpgeng/P1/model/'+ x)\n",
    "    model_id = x[:-4]\n",
    "    Cgeneid = []\n",
    "    for i in Cmodel.genes:\n",
    "        Cgeneid.append(i.id)\n",
    "        #Cgeneid = Cgeneid[:2]\n",
    "    Cseqs = {}\n",
    "    C_gene_mop = {}\n",
    "    C_gene_asso_reac_count = {}\n",
    "    C_protein_seqs = {}\n",
    "    \n",
    "    for i in tqdm(Cgeneid):\n",
    "        l = []\n",
    "        html = urllib.request.urlopen('http://bigg.ucsd.edu/models/'+ x[:-4] +'/genes/'+ i)\n",
    "        b = str(html.read()).split('\\\\n')\n",
    "        for j in range(len(b)):\n",
    "            if 'DNA Sequence' in b[j]:        \n",
    "                Cseqs[i] = re.findall(r'<p class=\"sequence\">(.+?)</p>',b[j+1])[0]\n",
    "            if '<h4>Strand:' in b[j]:        \n",
    "                C_gene_mop[i] = re.findall(r'<p>(.+?)</p>',b[j+1])[0]\n",
    "            if 'Protein Sequence' in b[j]:        \n",
    "                C_protein_seqs[i] = re.findall(r'<p class=\"sequence\">(.+?)</p>',b[j+1])[0]\n",
    "            if f'href=\"/models/{model_id}/reactions' in b[j]:\n",
    "                l.append(re.findall(r'/reactions/(.+?)\">',b[j])[0])\n",
    "        C_gene_asso_reac_count[i] = l\n",
    "    \n",
    "    C_ignored_genes = []\n",
    "    for i in Cgeneid:\n",
    "        if i not in Cseqs and i not in C_protein_seqs:\n",
    "            C_ignored_genes.append(i)\n",
    "        elif i not in C_protein_seqs and (len(Cseqs[i]) % 3 != 0 or len(Cseqs[i]) == 0):\n",
    "            C_ignored_genes.append(i)\n",
    "            \n",
    "    for i in C_ignored_genes:\n",
    "        Cgeneid.remove(i)\n",
    "    \n",
    "    # GC ratio\n",
    "    C_GC = {}\n",
    "    for i in Cgeneid:\n",
    "        gc = 0\n",
    "        for j in Cseqs[i]:\n",
    "            if j in ['G', 'C']:\n",
    "                gc += 1\n",
    "        C_GC[i] = round(gc / len(Cseqs[i]), 3)\n",
    "    print(len(C_GC)) \n",
    "\n",
    "    # associated reaction number\n",
    "    C_asso_rea_num = {}\n",
    "    for i in Cmodel.genes:\n",
    "        if i.id in C_ignored_genes: # exclude biomass function\n",
    "            continue\n",
    "        C_asso_rea_num[i.id] = len(i.reactions)\n",
    "    print(len(C_asso_rea_num))\n",
    "\n",
    "    # gene length\n",
    "    C_length = {}\n",
    "    for i in Cgeneid:\n",
    "        C_length[i] = len(Cseqs[i])\n",
    "    print(len(C_length))\n",
    "    \n",
    "    \n",
    "    C_rea_asso_enzy_count = {}\n",
    "    for i in tqdm(Cmodel.reactions, desc='crawling reactions info'):\n",
    "        rea_id = i.id if '_copy' not in i.id else i.id.split(\"_copy\")[0] # exclude copy reactions\n",
    "        html = urllib.request.urlopen(f'http://bigg.ucsd.edu/models/{model_id}/reactions/' + rea_id)\n",
    "        b = str(html.read()).split('\\\\n')\n",
    "        enzylis = []\n",
    "        for j in range(len(b)):\n",
    "            if '//identifiers.org/ec-code' in b[j]:        \n",
    "                enzylis.append( re.findall(r'identifiers.org/ec-code/(.+?)\"',b[j])[0].split(\".\")[0] )\n",
    "        C_rea_asso_enzy_count[rea_id] = enzylis\n",
    "        \n",
    "    C_gene_asso_enzy_count = []\n",
    "    for i in Cgeneid:\n",
    "        l = []\n",
    "        for j in C_gene_asso_reac_count[i]:\n",
    "            try:        \n",
    "                l += C_rea_asso_enzy_count[j]\n",
    "            except:\n",
    "                print(j)\n",
    "        C_gene_asso_enzy_count.append(l)\n",
    "    print(len(C_gene_asso_enzy_count))\n",
    "\n",
    "    for i in range(len(C_gene_asso_enzy_count)):\n",
    "        try:\n",
    "            C_gene_asso_enzy_count[i] = list(set(C_gene_asso_enzy_count[i]))\n",
    "        except:\n",
    "            print(C_gene_asso_enzy_count[i])\n",
    "\n",
    "    for i in range(len(C_gene_asso_enzy_count)):\n",
    "        for j in range(len(C_gene_asso_enzy_count[i])):\n",
    "            C_gene_asso_enzy_count[i][j] = int(C_gene_asso_enzy_count[i][j])\n",
    "\n",
    "    C_gene_asso_enzy = []\n",
    "    for i in C_gene_asso_enzy_count:\n",
    "        b = [1 if (k+1) in i else 0 for k in range(6)]\n",
    "        C_gene_asso_enzy.append(b)\n",
    "    print(len(C_gene_asso_enzy))\n",
    "\n",
    "    Cgc = []\n",
    "    for i in Cgeneid:\n",
    "        Cgc.append(C_GC[i])\n",
    "    print(len(Cgc))\n",
    "\n",
    "    Casn = []\n",
    "    for i in Cgeneid:\n",
    "        Casn.append(round(C_asso_rea_num[i]/10,3))\n",
    "    print(len(Casn))\n",
    "\n",
    "    Clen = []\n",
    "    for i in Cgeneid:\n",
    "        Clen.append(round(C_length[i]/100,3))\n",
    "    print(len(Clen))\n",
    "\n",
    "    Cmop = []\n",
    "    for i in Cgeneid:\n",
    "        if C_gene_mop[i] == 'Minus':\n",
    "            Cmop.append(-1)\n",
    "        elif C_gene_mop[i] == 'Plus':\n",
    "            Cmop.append(1)\n",
    "    print(len(Cmop))\n",
    "    \n",
    "    dic = {}\n",
    "    for i in range(len(Cgeneid)):\n",
    "        dic[Cgeneid[i]] = [Cmop[i], Clen[i], Casn[i], Cgc[i]] + C_gene_asso_enzy[i]\n",
    "\n",
    "    \n",
    "    with open(x[:-4]+'_10_features.pkl', 'wb') as f:\n",
    "        pickle.dump(dic, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5f92063",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e52bab6a",
   "metadata": {},
   "source": [
    "### Protein embedding "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd248a3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# first change path as the protein file cause path is different here\n",
    "\n",
    "from bio_embeddings.embed import ProtTransBertBFDEmbedder\n",
    "from Bio import SeqIO\n",
    "\n",
    "def data_prepare_embedding(model_id, middle_path):\n",
    "    \n",
    "    sequences = []\n",
    "    for record in SeqIO.parse(os.path.join(middle_path, f'{model_id}_protein_sequences.txt'), \"fasta\"):\n",
    "        sequences.append(record)\n",
    "\n",
    "    embedder = ProtTransBertBFDEmbedder()\n",
    "\n",
    "    embeddings = embedder.embed_many([str(s.seq) for s in sequences])\n",
    "\n",
    "    # `embed_many` returns a generator.\n",
    "    # We want to keep both RAW embeddings and reduced embeddings in memory.\n",
    "    # To do so, we simply turn the generator into a list!\n",
    "    # (this will start embedding the sequences!)\n",
    "\n",
    "    embeddings = list(embeddings)\n",
    "\n",
    "    reduced_embeddings = {s.id: list(ProtTransBertBFDEmbedder.reduce_per_protein(e)) for s,e in zip(sequences, embeddings)}\n",
    "\n",
    "    with open(os.path.join('/data1/xpgeng/P1/raw_data', f'{model_id}_protein_embedding.pkl'), \"wb\") as f:\n",
    "        pickle.dump(reduced_embeddings, f)\n",
    "        \n",
    "data_prepare_embedding('iML1515', '/data1/xpgeng/P1/raw_data/model_seq')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92f0a20a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
